### DISCUSSION TITLE
Streaming SBC or APTX over http to ffmpeg?

### DISCUSSION THREAD

**Discussion by mmcc-xx** [pschatzmann/arduino-audio-tools#825] [Q&A] [answered]
I'm trying to stream audio data captured via I2S over http. I'd like to do 44100hz sampling, so applying some compression seems like a good idea. I was able to make streams-i2s-webserver_aac work, but couldn't get the sampling rate over 22050hz. It was suggested that I try one of the codecs that Bluetooth uses.

That's where I am now. Starting with streams-i2s-webserver_aac as a model, I've tried both SBC and APTX encoding. I can connect to the server, the server seems to be serving up data, but I can't get ffmpeg to decode it. Ffmpeg runs a while then returns an error of "Invalid data found when processing input".

I have been able to get ffmpeg to translate a wav file to an SBC file and from SBC to a wav file.

My ffmpeg command looks something like this:

ffmpeg -i http://192.168.1.150 -dcodec sbc -t 15 output.wav

Any suggestions?

**Comment by pschatzmann** [comment, marked as answer]
Not sure what the issue is. ffmpeg does not give any meaningful hint...
Today I fixed the broken opus-ogg codec: maybe you can try [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-webserver/streams-generator-webserver_ogg/streams-generator-webserver_ogg.ino)

**Comment by mmcc-xx** [comment]
There's something odd going on...

Before trying this I updated the arduino-audio-tools library and cloned the libopus library for the first time.

With the example you suggested, when I try to open the stream with VLC, it take a long time to buffer up and the play back is very choppy.

When I try to save the stream to ogg, I get a bunch of errors, but it does produce a file. The output file was 35KB and could be opened with VLC. It contained the test tone but also some noise. I recorded for about 10 seconds, and the output file was about 6 seconds long.

Note that ffmpeg thinks the input stream has a sample rate of 48000Hz even though it has been configured to be 16000Hz. This is the same in every test.
```
me@myserver:~$ ffmpeg -i http://192.168.1.202 -t 15 test.ogg
ffmpeg version 4.4.2-0ubuntu0.22.04.1 Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 11 (Ubuntu 11.2.0-19ubuntu1)
  configuration: --prefix=/usr --extra-version=0ubuntu0.22.04.1 --toolchain=hardened --libdir=/usr/lib/x86_64-linux-gnu --incdir=/usr/include/x86_64-linux-gnu --arch=amd64 --enable-gpl --disable-stripping --enable-gnutls --enable-ladspa --enable-libaom --enable-libass --enable-libbluray --enable-libbs2b --enable-libcaca --enable-libcdio --enable-libcodec2 --enable-libdav1d --enable-libflite --enable-libfontconfig --enable-libfreetype --enable-libfribidi --enable-libgme --enable-libgsm --enable-libjack --enable-libmp3lame --enable-libmysofa --enable-libopenjpeg --enable-libopenmpt --enable-libopus --enable-libpulse --enable-librabbitmq --enable-librubberband --enable-libshine --enable-libsnappy --enable-libsoxr --enable-libspeex --enable-libsrt --enable-libssh --enable-libtheora --enable-libtwolame --enable-libvidstab --enable-libvorbis --enable-libvpx --enable-libwebp --enable-libx265 --enable-libxml2 --enable-libxvid --enable-libzimg --enable-libzmq --enable-libzvbi --enable-lv2 --enable-omx --enable-openal --enable-opencl --enable-opengl --enable-sdl2 --enable-pocketsphinx --enable-librsvg --enable-libmfx --enable-libdc1394 --enable-libdrm --enable-libiec61883 --enable-chromaprint --enable-frei0r --enable-libx264 --enable-shared
  libavutil      56. 70.100 / 56. 70.100
  libavcodec     58.134.100 / 58.134.100
  libavformat    58. 76.100 / 58. 76.100
  libavdevice    58. 13.100 / 58. 13.100
  libavfilter     7.110.100 /  7.110.100
  libswscale      5.  9.100 /  5.  9.100
  libswresample   3.  9.100 /  3.  9.100
  libpostproc    55.  9.100 / 55.  9.100
Input #0, ogg, from 'http://192.168.1.202':
  Duration: N/A, start: -0.009604, bitrate: N/A
  Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
Stream mapping:
  Stream #0:0 -> #0:0 (opus (native) -> vorbis (libvorbis))
Press [q] to stop, [?] for help
Output #0, ogg, to 'test.ogg':
  Metadata:
    encoder         : Lavf58.76.100
  Stream #0:0: Audio: vorbis, 48000 Hz, mono, fltp
    Metadata:
      encoder         : Lavc58.134.100 libvorbis
[libvorbis @ 0x55ef9e4afc80] Queue input is backward in time
    Last message repeated 15 times
[ogg @ 0x55ef9e513d00] Non-monotonous DTS in output stream 0:0; previous: -2754, current: -3128; changing to -2754. This may result in incorrect timestamps in the output file.
[libvorbis @ 0x55ef9e4afc80] Queue input is backward in time
    Last message repeated 1 times
(therse error messages continue ad infinitum)
```

When I try to save to wav, it seems to work but the clock timing the save is running really slowly. Note the "time" on the last line is listed as 0.60 seconds, even though it had been running for maybe 20 seconds. The output file sounded the same as the ogg file, but was of course much larger (1.2MB). It was also much longer in duration - VLC says 12 seconds.
```
me@myserver:~$ ffmpeg -i http://192.168.1.202 -t 15 test.wav
ffmpeg version 4.4.2-0ubuntu0.22.04.1 Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 11 (Ubuntu 11.2.0-19ubuntu1)
(repeated stuff deleted)
Input #0, ogg, from 'http://192.168.1.202':
  Duration: N/A, start: -0.009604, bitrate: N/A
  Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
Stream mapping:
  Stream #0:0 -> #0:0 (opus (native) -> pcm_s16le (native))
Press [q] to stop, [?] for help
Output #0, wav, to 'test.wav':
  Metadata:
    ISFT            : Lavf58.76.100
  Stream #0:0: Audio: pcm_s16le ([1][0][0][0] / 0x0001), 48000 Hz, mono, s16, 768 kb/s
    Metadata:
      encoder         : Lavc58.134.100 pcm_s16le
size=    2048kB time=00:00:00.60 bitrate=27764.4kbits/s speed=0.0315x
```

I also tried copying to an opus file without reencoding. The resulting file was very 36KB. It also contained the test tone but was only about 3 seconds long.
```
me@myserver:~$ ffmpeg -i http://192.168.1.202 -t 15 -c copy test.opus
ffmpeg version 4.4.2-0ubuntu0.22.04.1 Copyright (c) 2000-2021 the FFmpeg developers
  built with gcc 11 (Ubuntu 11.2.0-19ubuntu1)
(repeated stuff deleted)
Input #0, ogg, from 'http://192.168.1.202':
  Duration: N/A, start: -0.009604, bitrate: N/A
  Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
Output #0, opus, to 'test.opus':
  Metadata:
    encoder         : Lavf58.76.100
  Stream #0:0: Audio: opus, 48000 Hz, mono, fltp
    Metadata:
      encoder         : Lavf58.76.100
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
Press [q] to stop, [?] for help
size=     170kB time=00:00:01.74 bitrate= 795.6kbits/s speed=0.0339x
```

Also, between tests I had to restart the ESP32. If I didn't, I'd get an error like...
`[ogg @ 0x564e013e1680] Codec not found`

The AAC example streams-generator-server_aac.ino works perfectly.
The corresponding MP3 example throws this error in the serial monitor after download:
```
[Error] c:\Users\Me\Documents\Arduino\libraries\arduino-liblame\src\liblame\lame.c : 2772 - calloc(1,85840) -> 0x0 
available MALLOC_CAP_8BIT: 110580 / MALLOC_CAP_32BIT: 110580  / MALLOC_CAP_SPIRAM: 0
```
